src/stdlib_selection.fypp

```diff
@@ -0,0 +1,276 @@
+              !! If we know that:
+              !!    the k-th smallest entry of `a` is in `a(left:right)`
+              !! and also that:
+              !!    `maxval(a(1:(left-1))) <= minval(a(left:right))`
+              !! and:
+              !!    `maxval(a(left:right))) <= minval(a((right+1):size(a)))`
+              !! then one or both bounds can be specified to narrow the search.
+              !! The constraints are available if we have previously called the
+              !! subroutine with different `k` (because of how `a(:)` becomes
+              !! partially sorted, see documentation for `a(:)`).
+
+          ${inttype}$ :: l, r, mid, iPivot
+          integer, parameter :: ip = ${intkind}$
+
+          l = 1_ip
+          if(present(left)) l = left
+          r = size(a, kind=ip)
+          if(present(right)) r = right
+
+          if(k < 1_ip .or. k > size(a, kind=ip) .or. l > r .or. l < 1_ip .or. &
+              r > size(a, kind=ip)) then
+              error stop "select must have 1 <= k <= size(a), and 1 <= left <= right <= size(a)";
+          end if
+
+          do
+              mid = l + ((r-l)/2_ip) ! Avoid (l+r)/2 which can cause overflow
+
+              call medianOf3(a, l, mid, r)
+              call swap(a(l), a(mid))
+              call partition(a, l, r, iPivot)
+
+              if (iPivot < k) then
+                l = iPivot + 1_ip
+              elseif (iPivot > k) then
+                r = iPivot - 1_ip
+              elseif (iPivot == k) then
+                kth_smallest = a(k)
+                return
+              end if
+          end do
+
+          contains
+              pure subroutine swap(a, b)
+                  ${arraytype}$, intent(inout) :: a, b
+                  ${arraytype}$ :: tmp
+                  tmp = a; a = b; b = tmp
+              end subroutine
+
+              pure subroutine medianOf3(a, left, mid, right)
+                  ${arraytype}$, intent(inout) :: a(:)
+                  ${inttype}$, intent(in) :: left, mid, right 
+                  if(a(right) < a(left)) call swap(a(right), a(left))
+                  if(a(mid)   < a(left)) call swap(a(mid)  , a(left))
+                  if(a(right) < a(mid) ) call swap(a(mid)  , a(right))
+              end subroutine
+
+              pure subroutine partition(array,left,right,iPivot)
+                  ${arraytype}$, intent(inout) :: array(:)
+                  ${inttype}$, intent(in) :: left, right
+                  ${inttype}$, intent(out) :: iPivot 
+                  ${inttype}$ :: lo,hi
+                  ${arraytype}$ :: pivot
+
+                  pivot = array(left)
+                  lo = left
+                  hi=right
+                  do while (lo <= hi)
+                    do while (array(hi) > pivot)
+                      hi=hi-1_ip
+                    end do
+                    do while (lo <= hi )
+                       if(array(lo) > pivot) exit
+                      lo=lo+1_ip
+                    end do
```

Comment:
This exit is nested within two looping constructs. A named loop should help make this unambiguous. Same goes for other places that use an exit.

Reply:
Good point -- addressed in a commit just now